org.mixedrealitytoolkit.uxcore/Experimental/NonNativeKeyboard/NonNativeKeyTouchAdapter.cs

```diff
@@ -0,0 +1,145 @@
+            buttonCollider.size = size;
+            buttonColliderDefaultCenter = new Vector3((size.x + ColliderMargin) / 2.0f,
+                (-size.y - ColliderMargin) / 2.0f, ColliderZDelta);
+            buttonCollider.center = buttonColliderDefaultCenter;
+            
+            image = GetComponent<Graphic>();
+            defaultImageColor = image.color;
+            var button = GetComponent<Button>();
+            
+            interactable = gameObject.EnsureComponent<StatefulInteractable>();
+            interactable.firstSelectEntered.AddListener(selectArgs =>
+            {
+                if (selectArgs.interactorObject is not PokeInteractor ||
+                    Time.time - lastClickTime < ReClickDelayTime)
+                {
+                    return;
+                }
+
+                button.onClick.Invoke();
+                StartCoroutine(MoveButton(defaultPosition, animatedPosition));
+            });
+            button.interactable = false;
+
+            interactable.lastSelectExited.AddListener(_ =>
+            {
+                StartCoroutine(MoveButton(animatedPosition, defaultPosition));
+            });
+
+            interactable.firstHoverEntered.AddListener(hoverArgs =>
+            {
+                SetColorOnHoverPoke(hoverArgs.interactorObject, button.colors.highlightedColor);
+            });
+            
+            interactable.lastHoverExited.AddListener(hoverArgs =>
+            {
+                SetColorOnHoverPoke(hoverArgs.interactorObject, defaultImageColor);
+            });
+        }
+
+        private void SetColorOnHoverPoke(IXRHoverInteractor interaction, Color color)
+        {
+            if (interaction is PokeInteractor)
+            {
+                image.color = color;
+            }
+        }
+        
+        private void Update()
+        {
+            // fallback for if the user very rapidly moves their finger over the keyboard and
+            // interactable.lastHoverExited does not get fired (or too early)
+            if (interactable.HoveringPokeInteractors.Count == 0 && image.color != defaultImageColor)
+            {
+                image.color = defaultImageColor;
+            }
+        }
+
+        private IEnumerator MoveButton(Vector3 startPos, Vector3 endPos)
+        {
+            if (transform.localPosition == endPos)
+            {
+                yield break;
+            }
+            const float rate = 1.0f / AnimationTime;
+            var i = 0.0f;
+            while (i < 1.0f)
+            {
+                i += Time.deltaTime * rate;
+                var newPos = Vector3.Lerp(startPos, endPos, Mathf.SmoothStep(0f, 1f, i));
+                transform.localPosition = newPos; buttonCollider.center = buttonColliderDefaultCenter - (newPos - defaultPosition);
+                yield return null;
+            }
+        }
+
+        private void OnDestroy()
+        {
+            interactable.hoverExited.RemoveAllListeners();
```

Comment:
Please only remove the explicit handlers you added.

Reply:
Refactored anonymous lambda's into separate methods that can be individually removed. However - this _is_ the OnDestroy method, right? Nothing much happens after this ;)